theano/compile/builders.py

```diff
@@ -216,64 +256,94 @@ def __str__(self):
         return '%(name)s{inline=%(is_inline)s}' % locals()
 
     def _recompute_grad_op(self):
-        if isinstance(self._grad_op, OpFromGraph):
+        local_inputs = self.local_inputs
+        local_outputs = self.local_outputs
+        inp_len = len(local_inputs)
+        grad_op = self._grad_op
+
+        if isinstance(grad_op, OpFromGraph):
             self._grad_op_is_cached = True
+            self._grad_op_overrides_l = [None] * len(self.local_inputs)
             return
-        output_grads = [out_t() for out_t in self.output_types]
-        if self._grad_op is None:
-            self._grad_op = []
 
-        # we need to convert a list/function into an OfG instance
-        if isinstance(self._grad_op, list):
+        output_grads = [out_t() for out_t in self.output_types]
+        TYPE_ERR_MSG = 'Gradient override should be (single or list of)' \
+            'OpFromGraph | Ellipsis | None | 0 | callable, got %s'
+        # we need to convert _grad_op into an OfG instance
+        if grad_op is Ellipsis:
+            self._grad_op_tflags = bytes(inp_len)
+            all_grads_l = theano.gradient.grad(
+                cost=None,
+                known_grads=OrderedDict(izip(local_outputs, output_grads)),
+                wrt=local_inputs,
+                disconnected_inputs='ignore')
+            all_grads_ov_l = [None] * inp_len
+        elif grad_op is None:
+            all_grads_l = [inp.zeros_like() for inp in local_inputs]
+            all_grads_ov_l = [self.ofg_null_t()] * inp_len
+        elif grad_op is 0:
```

Comment:
I'm not sure that 0 is guaranted to be a singleton by all Python distribution. Unless you know why it is competly safe, I would prefer this safer version `type(grad_op) is int and grad_op==0`